### PR TITLE
Fix build errors on djgpp

### PR DESCRIPTION
**Pull request by jwt27**
I tried cross-compiling the current `master` branch for djgpp and ran into some
errors.  This patch set aims to solve those.

On each commit I specified `CLA: trivial`, since they are small tweaks and do
not introduce any new features.  Only build errors are fixed.

Remaining issues:
* `test/rsa_complex.c` does not compile due to missing header `<complex.h>`.
* Linking is only possible when compiled with `-march=i486` or better, due to
  missing atomic instructions on i386 (these are compiled to libatomic calls,
  which are also missing).  Is it safe to define `-DOPENSSL_DEV_NO_ATOMICS`
  here, given that we have no (native) pre-emptive threads?

### CODE REVIEW

**Review by levitte — DISMISSED** [review]

**Comment by levitte** [comment]
I marked this for the 3.0 branch, 'cause I think of this as bug fixes and that they should thus be backported. I'm unsure, though, if this applies cleanly on that branch, it's possible that a separate PR is needed.

It may be that the same reasoning should be made for 1.1.1.

Is this something you're willing to look into, @jwt27?

**Comment by mattcaswell** [comment]
> It may be that the same reasoning should be made for 1.1.1.

1.1.1 is receiving security fixes only - so this would not be backported there.

**Review by t8m — COMMENTED** [review]
Unfortunately at least the last commit is outside of what we would accept with CLA: trivial. Can you please sign the regular CLA? https://www.openssl.org/policies/cla.html

**Comment by jwt27** [comment]
> I marked this for the 3.0 branch, 'cause I think of this as bug fixes and that they should thus be backported. I'm unsure, though, if this applies cleanly on that branch, it's possible that a separate PR is needed.
> 
> It may be that the same reasoning should be made for 1.1.1.
> 
> Is this something you're willing to look into, @jwt27?

Sure, I had a look:

On `openssl-3.0`, the first and third patches apply cleanly.  The other two are
not needed.

On `OpenSSL_1_1_1-stable`, only the third patch applies.  `AR`/`ARFLAGS` are
correctly set on configure, however `-lz` is missing from the linker flags.
Inheriting config from `BASE_unix` does resolve this, but the respective patch
would need to be edited.

The remaining issues listed in my first post also apply to these branches.

> Unfortunately at least the last commit is outside of what we would accept with CLA: trivial. Can you please sign the regular CLA? https://www.openssl.org/policies/cla.html

Alright, will do.  I just realized using `gmtime_r()` produces the wrong
result, so I'll have to rewrite that patch anyway.

File: crypto/asn1/a_time.c [hunk 1]
```diff
@@ -627,17 +627,38 @@ time_t ossl_asn1_string_to_time_t(const char *asn1_string)
...
+    }
+#elif defined(USE_TIMEGM)
     timestamp_utc = timegm(timestamp_tm);
-    OPENSSL_free(timestamp_tm);
 #else
-    timestamp_local = mktime(timestamp_tm);
-    OPENSSL_free(timestamp_tm);
-
-    timestamp_utc = timestamp_local - timezone;
+    tzset();
```
Comment (t8m) [inline crypto/asn1/a_time.c:658]: This causes Windows CI build failure. Is it actually needed? mktime() should call it implicitly. This is outside of DJGPP support so better not add it.

Comment (jwt27) [reply crypto/asn1/a_time.c:658]: It seemed a bit pointless to use a separate variable here and do it in two steps.  Then I realized `timezone` may not even be set until the call to `mktime()`, so doing both in one expression would be undefined behavior.  I added `tzset()` then which I see now doesn't work on Windows.

**Comment by jwt27** [comment]
CLA is signed, so this PR is ready for review.  I was told I had to close and
reopen it to remove a "needs CLA" tag, but it was never added so I suppose
that's not necessary.

**Review by t8m — APPROVED** [review]

**Review by levitte — APPROVED** [review]

**Comment by jwt27** [comment]
Another question - would it be useful to add djgpp as a CI build target?  I
maintain an Ubuntu PPA specifically for that purpose.  In fact, I ran into
these build errors as I was planning to add an OpenSSL package to it.

If there is any interest, and there are no objections to adding a dependency on
some random person's PPA, I can submit a PR to add it in `cross-compiles.yml`.

**Comment by t8m** [comment]
Merged to master branch. Cherry-picked to 3.0 branch except for the last commit which does not apply there.

Closing.


> If there is any interest, and there are no objections to adding a dependency on some random person's PPA, I can submit a PR to add it in `cross-compiles.yml`.

I am not sure about that. You can submit a PR if you want but this is something we would need to discuss within OTC IMO before merging it.

**Comment by jwt27** [comment]
Thanks for merging!  I see now `e_os.h` is in a different location on the 3.0 branch, didn't notice that at first.

> > If there is any interest, and there are no objections to adding a dependency on some random person's PPA, I can submit a PR to add it in `cross-compiles.yml`.
> 
> I am not sure about that. You can submit a PR if you want but this is something we would need to discuss within OTC IMO before merging it.

That is understandable.  I'll submit the PR then and see what happens.

**Pull request closed without merging**